services/web/server/src/simcore_service_webserver/projects/projects_api.py

```diff
@@ -184,19 +189,65 @@ async def remove_services_and_data():
 ## PROJECT NODES -----------------------------------------------------
 
 
-@observe(event="SIGNAL_PROJECT_CLOSE")
+async def retrieve_and_notify_project_locked_state(
+    user_id: int, project_uuid: str, app: web.Application
+):
+    project = await get_project_for_user(app, project_uuid, user_id, include_state=True)
+    await notify_project_state_update(app, project)
+
+
 async def remove_project_interactive_services(
-    user_id: Optional[int], project_uuid: Optional[str], app: web.Application
+    user_id: int, project_uuid: str, app: web.Application, notify_users: bool = True
 ) -> None:
-    # save the state if the user is not a guest. if we do not know we save in any case.
-    with suppress(director_exceptions.DirectorException):
-        # here director exceptions are suppressed. in case the service is not found to preserve old behavior
-        await director_api.stop_services(
-            app=app,
-            user_id=user_id,
-            project_id=project_uuid,
-            save_state=not await is_user_guest(app, user_id) if user_id else True,
+    # NOTE: during the closing process, which might take awhile,
+    # the project is locked so no one opens it at the same time
+    try:
+        log.debug(
+            "removing project interactive services for project [%s] and user [%s]",
+            project_uuid,
+            user_id,
         )
+        async with await lock_project(
+            app,
+            project_uuid,
+            ProjectStatus.CLOSING,
+            user_id,
+            await get_user_name(app, user_id),
+        ):
+            if notify_users:
+                await retrieve_and_notify_project_locked_state(
+                    user_id, project_uuid, app
+                )
+
+            # save the state if the user is not a guest. if we do not know we save in any case.
+            with suppress(director_exceptions.DirectorException):
+                # here director exceptions are suppressed. in case the service is not found to preserve old behavior
+                await director_api.stop_services(
+                    app=app,
+                    user_id=user_id,
```

Comment:
use pylance and check all the type mismatches.. ``user_id`` being ``int`` or ``str`` is a classic. We have it in so many places ... we should slowly get it right.

Reply:
I agree with you but this will be replaced by @GitHK PR on the dynamic sidecar.